---
title: "Dynamic_NicheNet_test"
author: "Heesoo Song"
date: '2021 10 28 '
output: html_document
---

# 0. Install/Load packages

```{r}
#install.packages("devtools")
#install.packages("htmltools")
#BiocManager::install("limma")
#devtools::install_github("saeyslab/nichenetr")
#install.packages("tidyverse")
#BiocManager::install("muscat")
```


```{r, verbose = FALSE}
library(SeuratObject)
library(Seurat)
library(nichenetr)
library(tidyverse)
library("ggplot2")
library(muscat)
```

# 1. Load datasets

**Prior model**
Numbers in the ligand-target matrix represents the ligand-target regulatory potential scores between all pairs of ligands and target genes. These scores are calculated using network propagation methods on the integrated networks to propagate the signal from a ligand, over receptors, signaling proteins and transcriptional regulators, to end at target genes.
```{r, cache = TRUE}
# Load ligand-target prior model
ligand_target_matrix = readRDS(url("https://zenodo.org/record/3260758/files/ligand_target_matrix.rds"))
ligand_target_matrix[1:5, 1:5]
```

**Expression data**
```{r, cache = TRUE}
# Load processed expression data of interacting cells
hepatectomy_expression = readRDS("C:/Users/pc/Desktop/MastersProject/Dynamic_NicheNet/seurat_obj_phx_25102021.rds")

# Total 91021 cells, with 5000 genes
```
- percent.mt = calculates the percentage of counts originating from mitochondrial genes - high percentage mitochondrial genes can be a sign of low-quality cells 
- S.Score =	score related to the expression of genes specific for cells in the S phase of the cell cycle
- G2M.Score	= score related to the expression of genes specific for cells in the G2 and M phases of the cell cycle
- Phase	cell = cycle phase (based on the G2M, S and G1 scores)
- phase	cell = cycle phase (based on the scores above)
- G1.Score =	score related to the expression of genes specific for cells in the S (G1?) phase of the cell cycle
- retained_nFeature_RNA	= indicates whether the cells is retained (TRUE) or removed (FALSE) based on the number of unique features/genes expressed in that cell. Cells with too few or too many features are sometimes removed because this can point to low-quality cells (too few features)  or doublets (too many unique features)
- retained_nCount_RNA	= indicates whether the cells is retained (TRUE) or removed (FALSE) based on the number of counts in that cell. Cells with too few or too many counts are sometimes removed because this can point to low-quality cells (few counts) or doublets (too many counts)
- retained_percent.mt	= indicates whether the cells is retained (TRUE) or removed (FALSE) based on the percent.mt -- if too high: often low-quality (exception: very metabolic active cells like...hepatocytes!)
- retained	= indicates whether the cells is retained (TRUE) or removed (FALSE) based on the criteria mentioned above
- RNA_snn_res.1	= cluster identity of the cell (clusterin was performed with resolution = 1)
- seurat_clusters	= cluster identity of the cell (here the same as RNA_snn_res.1, because no other clustering resolutions were tried out)

# 2. Explore data

## 1) Metadata exploration

### a. Check quality of the cells
From the description of the data objects, outlying cells are removed since they often indicates the low-quality. Thus, check which cells are retained in three conditions.
```{r}
retained_nFeature <- hepatectomy_expression$retained_nFeature_RNA
retained_nCount <- hepatectomy_expression$retained_nCount_RNA
retained_mt <- hepatectomy_expression$retained_percent.mt

df_retained <- as.data.frame(cbind(retained_nFeature, retained_nCount, retained_mt))
```

```{r}
retained_three <- df_retained[df_retained$retained_nFeature == TRUE & df_retained$retained_nCount == TRUE & df_retained$retained_mt == TRUE,]
```
--> 91021 cells are retained in all three conditions. Turns out that all cells are in good quality.

### b. Check cell-cycle phases

Phase and phase?? Robin also doesn't know. Maybe thresholds were different in differentiating phases. Let it be for now. Define it in detail if the cell-cycle phase matters later on.
```{r}
hepatectomy_expression$Phase[1:10]
hepatectomy_expression$phase[1:10]
```

### c. Check cell-types

First, have a look at frequencies for all existing cell-types.

```{r}
library(plyr)
celltype_level <- levels(factor(hepatectomy_expression$celltype))
celltype_level
celltype_freq <- count(hepatectomy_expression$celltype)

bp_celltype <- barplot(celltype_freq$freq, ylim = c(0, 30000),names.arg = celltype_freq$x, ylab="Counts", main = "Cell-type Frequency Table", las=2)
text(bp_celltype, celltype_freq$freq, labels=celltype_freq$freq, pos=3, xpd=TRUE)
```

cell types of interest in this study are:  c("cholangiocyte","endothelial","hepatocyte","kupffer","monocyte","neutrophil","stellate")

```{r}
celltype_OI <- c("cholangiocyte","endothelial","hepatocyte","kupffer","monocyte","neutrophil","stellate")


# Proportion of cell-types of interest
total_counts <- sum(celltype_freq$freq)
total_counts

df_celltype_OI <- celltype_freq[celltype_freq$x %in% celltype_OI,]
celltype_OI_counts <- sum(df_celltype_OI$freq)
celltype_OI_counts

ratio_cOI.total <- celltype_OI_counts / total_counts * 100
ratio_cOI.total
```
**89.81%** of the total cells are in cell-types of interest.

Then, look at the compartment of each type in the cell-types of interest.
```{r}
# Compartment of each type in the cell-types of interest
df_celltype_OI$ratio <- with(df_celltype_OI, freq / celltype_OI_counts * 100)
df_celltype_OI
```

### d. Check cells per experiment (time points)

**factor levels in time point variables**
```{r}
levels(factor(hepatectomy_expression$timepoint)) # 2D & 48h
levels(factor(hepatectomy_expression$time_experiment))
levels(factor(hepatectomy_expression$time))
levels(factor(hepatectomy_expression$experiment))
```
- What is phase_1 & phase_2? 
- There are only 0_phase_1 (ctr) and 48_phase_1 for phase_1 experiments, and only 48h overlaps in phase_1 experiment and phase_2 experiment. 
- How can we make use of this? Compare two phases at 48h, control noises, and finally use 0_phase_1 as control of whole experiment?
- What is S3 and S4 in sample name? e.g.AACACACCACCCTCTA.FRS4_48h_S4, TTTGACTCACCTATCC.FRS3_ctr_S3

**Analyze number of cell(type)s per time-point**
```{r}
# create dataframe for analysis
df_time <- as.data.frame(cbind(hepatectomy_expression$time, hepatectomy_expression$celltype))
colnames(df_time) <- c("time", "celltype")

df_time$time_type <- paste0(df_time$time, "_", df_time$celltype) # grouping. 11x7=77 factors
df_time <- df_time[df_time$celltype %in% celltype_OI,]
head(df_time)


# Timepoint frequency table
time_type.freq <- count(df_time$time_type)
time_type.freq$time <- sapply(strsplit(time_type.freq$x, "_"), "[", 1)
time_type.freq$type <- sapply(strsplit(time_type.freq$x, "_"), "[", 2)
head(time_type.freq)
```

**Total number of cells in each timepoint**
```{r}
aggregate(time_type.freq$freq, by=list(Category=time_type.freq$time), FUN=sum)
```

**Stacked cell-type frequency plot in each timepoint**
```{r}
# Plot frequencies in stacked bar plot
time_type.freq$time <- factor(time_type.freq$time, levels = levels(factor(hepatectomy_expression$time)))

time_type.plot <- ggplot(time_type.freq, aes(x = time, y = freq, fill = type)) + 
  geom_bar(stat = "identity") +
  geom_text(aes(label = freq), position = position_stack(vjust = 0.5), size = 3) +
  ggtitle("Cell-type frequencies in each timepoint")

time_type.plot
```

**Percent stack cell-type frequency plot in each timepoint**
```{r}
library(dplyr)
# calculate percentage
dt <- ddply(time_type.freq, .(time), transform, percent = freq/sum(freq)) 

# Plot frequencies in percent-stacked bar plot
time_type.percent.plot <- ggplot(dt, aes(x = time, y = freq, fill = type)) + 
  geom_bar(position = "fill", stat = "identity") +
  ggtitle("Cell-type percent frequencies in each timepoint") +
  geom_text(aes(label = paste0(sprintf("%1.1f", percent*100), "%")),
            position = position_fill(vjust=0.5), size = 3)

time_type.percent.plot
```
From the total number of cells in each time point and "Cell-type frequencies in each timepoint" plot, you can see that the number of cells in 48h outstands the other timepoints. This can be explained because there were two experiments conducted at 48h, "48_phase_1" & "48_phase_2".

From the "Cell-type percent frequencies in each timt point" plot, you can observe that the proportion of cell-types differ in each time points.


**Compare 48h_phase_1 & 48h_phase_2**
```{r}
df_48h <- as.data.frame(cbind(hepatectomy_expression$time_experiment, hepatectomy_expression$experiment, hepatectomy_expression$celltype))

colnames(df_48h) <- c("time_experiment", "experiment", "celltype")

df_48h <- df_48h[df_48h$time_experiment %in% c("48_phase_1", "48_phase_2") & df_48h$celltype %in% celltype_OI,]

df_48h$phase_type <- paste0(df_48h$time_experiment, ".", df_48h$celltype)
df_48h.freq <- count(df_48h$phase_type)

df_48h.freq$phase <- sapply(strsplit(df_48h.freq$x, "\\."), "[", 1)
df_48h.freq$type <- sapply(strsplit(df_48h.freq$x, "\\."), "[", 2)
head(df_48h.freq)
```
```{r}
aggregate(df_48h.freq$freq, by=list(Category=df_48h.freq$phase), FUN=sum)
```

```{r}
df_48h.plot <- ggplot(df_48h.freq, aes(x = phase, y = freq, fill = type)) + 
  geom_bar(stat = "identity") +
  geom_text(aes(label = freq), position = position_stack(vjust = 0.5), size = 3) +
  ggtitle("Cell-type frequencies at 48h")

df_48h.plot
```

```{r}
# calculate percentage
dt_48h <- ddply(df_48h.freq, .(phase), transform, percent = freq/sum(freq)) 

# Plot frequencies in percent-stacked bar plot
df_48h.percent.plot <- ggplot(dt_48h, aes(x = phase, y = freq, fill = type)) + 
  geom_bar(position = "fill", stat = "identity") +
  ggtitle("Cell-type percent frequencies at 48h") +
  geom_text(aes(label = paste0(sprintf("%1.1f", percent*100), "%")),
            position = position_fill(vjust=0.5), size = 3)

df_48h.percent.plot
```
Whole number of cells in phase 1 and phase 2 are almost the same, 8851 cells and 8499 cells respectively. However, cell-type composition of phase 1 and phase 2 differs quite a lot. 
Phase 1: endothelial > hepatocyte > kupffer > neutrophil > monocyte > cholangiocyte > stellate
Phase 2: endothelial > kupffer > stellate > hepatocyte > monocyte > neutrophil > cholangiocyte
  - What would have cause this? 
  - How significant this difference is? 
  - Which one should we keep? 
  - Would batch effect have caused this? 
  - Should we fix this differences? 
  - What about the control which is also another phase 1? Should we also fix it to use it as a control of whole experiments?


## 2) Expression data exploration

### a. Expression for whole dataset

```{r}
expression = hepatectomy_expression$nCount_RNA
head(expression)
```
**UMAP: cell-type clusters**
```{r}
Command(hepatectomy_expression, "FindVariableFeatures.RNA")
```

```{r}
Command(hepatectomy_expression, "RunUMAP.RNA.pca")
```

```{r}
DimPlot(hepatectomy_expression)
```
```{r}
#pbmc <- FindVariableFeatures(hepatectomy_expression, selection.method = "vst", nfeatures = 2000) # This step already done in given object

# Identify the 10 most highly variable genes (features)
top10 <- head(VariableFeatures(hepatectomy_expression), 10)
top10
```

```{r}
#jpeg("variableFeatures.jpg", width = 2000, height = 1000)
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(hepatectomy_expression)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = FALSE)
plot2
#dev.off()
```

```{r}
# cutoff in quantile [0.10, 0.90]
FeaturePlot(hepatectomy_expression, features = "ENSMUSG00000017002", min.cutoff = "q10", max.cutoff = "q90", label = TRUE) #"ENSMUSG00000040026"
```

```{r}
# The expression level of top10 features in each cell-type
DotPlot(hepatectomy_expression, features = top10) + RotatedAxis()
```


### b. Expressions per timepoint

```{r}
levels(Idents(object=hepatectomy_expression))
```

```{r}
hepa_celltypeOI = subset(x = hepatectomy_expression, idents = celltype_OI)
```

```{r}
DimPlot(hepa_celltypeOI)
```


```{r}
hepa_t0 <- subset(x = hepa_celltypeOI, subset = time_experiment == "0_phase_1")
hepa_t3 <- subset(x = hepa_celltypeOI, subset = time_experiment == "3_phase_2")
hepa_t6 <- subset(x = hepa_celltypeOI, subset = time_experiment == "6_phase_2")
hepa_t12 <- subset(x = hepa_celltypeOI, subset = time_experiment == "12_phase_2")
hepa_t24 <- subset(x = hepa_celltypeOI, subset = time_experiment == "24_phase_2")
hepa_t36 <- subset(x = hepa_celltypeOI, subset = time_experiment == "36_phase_2")
hepa_t48_p1 <- subset(x = hepa_celltypeOI, subset = time_experiment == "48_phase_1")
hepa_t48_p2 <- subset(x = hepa_celltypeOI, subset = time_experiment == "48_phase_2")
hepa_t72 <- subset(x = hepa_celltypeOI, subset = time_experiment == "72_phase_2")
hepa_t96 <- subset(x = hepa_celltypeOI, subset = time_experiment == "96_phase_2")
hepa_t120 <- subset(x = hepa_celltypeOI, subset = time_experiment == "120_phase_2")
hepa_t168 <- subset(x = hepa_celltypeOI, subset = time_experiment == "168_phase_2")
```

```{r}
subsets <- c(hepa_t0, hepa_t3, hepa_t6, hepa_t12, hepa_t24, hepa_t36, hepa_t48_p1, hepa_t48_p2, hepa_t72, hepa_t96, hepa_t120, hepa_t168)
experiments <- c("0_phase_1", "3_phase_2", "6_phase_2", "12_phase_2", "24_phase_2", "36_phase_2", "48_phase_1", "48_phase_2", "72_phase_2", "96_phase_2", "120_phase_2", "168_phase_2")
```

**Change in cluster through time**
```{r}
for (i in 1:12){
  baseplot <- DimPlot(subsets[[i]])
  plotwithtitle <- baseplot + labs(title = experiments[i])
  print(plotwithtitle)
}
```
Cluster shape changes through time. It is again notable the difference between 48h_phase1 and 48h_phase2, which is supposed to be similar. 


**Change in expression level through time in a feature**
```{r}
for (i in 1:12){
  feature_baseplot <- FeaturePlot(subsets[[i]], features = "ENSMUSG00000017002", min.cutoff = 1, max.cutoff = 6 ,label = TRUE)
  featureplotwithtitle <- feature_baseplot + labs(title = paste0("ENSMUSG00000017002: ",experiments[i]))
  print(featureplotwithtitle)
}
```

- Nothing really special from the analysis above. Everything was well clustered, and showed enough changes through time
- What to do next? What will be more interesting to look at?
  - Changes in each single cell?
- 


**Timepoint as features**
```{r}
DimPlot(hepa_celltypeOI, group.by="time_experiment")
```



**Cell-cycle as features**
```{r}
#hepa_cellcycle <- RunPCA(hepa_cellcycle, features = c("S.Score", "G2M.Score"))
DimPlot(hepa_celltypeOI, group.by= "Phase")
DimPlot(hepa_celltypeOI, group.by= "phase")
```

# 3. Differentially Expressed Genes

## 1) Seurat Function: FindAllMarkers()
Matrix containing a ranked list of putative markers, and associated statistics (p-values, ROC score, etc.).
Default:
  - Wiocoxon test
  - Pseudocount.use = 1: Pseudocount to add to averaged expression values when calculating logFC
```{r}
# all.markers <- FindAllMarkers(hepa_celltypeOI, slot = "counts")
# DE_FindAllMarkers.png

#write.csv(all.markers,"C:/Users/pc/Desktop/MastersProject/Dynamic_NicheNet/DE_FindAllMarkers.csv")
```


## 2) Pseudobulk data DE analysis

Different ways to create pseudobulk matrix
- AggregateExpression(): Returns aggregated (summed) expression values for each identity class
- Muscat
- Manual creation: https://hbctraining.github.io/scRNA-seq_online/lessons/pseudobulk_DESeq2_scrnaseq.html


Muscat required change of object type into *SingleCellExperiment* but it kept rising error. Thus went for manual creation of matrix (3rd option). You can refer to *DE_EdgeR.Rmd*.

```{r}
# Extract the rlog matrix from the object and compute pairwise correlation values
rld_mat <- assay(rld)
rld_cor <- cor(rld_mat)

# Plot heatmap
pheatmap(rld_cor, annotation = cluster_metadata[, c("cluster_id"), drop=F])
```

# 3. Define expressed genes

# 4. Define gene set of interest and background of genes

# 5. Define a set of potential ligands

# 6. Perform NicheNet's ligand activity analysis on the gene set of interest

# 7. Infer target genes of top-ranked ligands and visualize in a heatmap

# 8. Validation



