---
title: "Dynamic_NicheNet"
author: "Heesoo Song"
date: '2022 2 21 '
output: html_document
---

# 0. Install/Load packages

```{r}
#install.packages("devtools")
#install.packages("htmltools")
#BiocManager::install("limma")
#devtools::install_github("saeyslab/nichenetr")
#install.packages("tidyverse")
```


```{r}
library(nichenetr)
library(tidyverse)
library(Matrix)
library(dplyr)
```

# 1. Load datasets

**Prior model**
Numbers in the ligand-target matrix represents the ligand-target regulatory potential scores between all pairs of ligands and target genes. These scores are calculated using network propagation methods on the integrated networks to propagate the signal from a ligand, over receptors, signaling proteins and transcriptional regulators, to end at target genes.
```{r}
# Load ligand-target prior model
ligand_target_matrix = readRDS(url("https://zenodo.org/record/5884439/files/ligand_target_matrix_nsga2r_final_mouse.rds"))  # New version
#ligand_target_matrix = readRDS(url("https://zenodo.org/record/3260758/files/ligand_target_matrix.rds"))  # Old version
ligand_target_matrix[1:5, 1:5]
```

```{r}
# Load processed expression data of interacting cells
hepa_rds <- readRDS("C:/Users/pc/Desktop/MastersProject/Dynamic_NicheNet/seurat_obj_phx_25102021.rds")
hepa_cluster <- read.csv("C:/Users/pc/Desktop/MastersProject/Dynamic_NicheNet/DE_pseudobulk_cluster.csv")
head(hepa_cluster)
```

```{r}
# Preprocess cluster data
## Assign clear rownames
rownames(hepa_cluster) <- NULL
hepa_cluster <- hepa_cluster %>% column_to_rownames(var="Row.names")

## Add celltype & Symbol columns to cluster matrix
extract_geneID <- as.data.frame(str_split(rownames(hepa_cluster), "_", simplify=TRUE))
hepa_cluster$Symbol <- extract_geneID$V1
hepa_cluster$celltype <- extract_geneID$V2
head(hepa_cluster)
```


```{r}
# Preprocess expression data
## Subset to seven cell-types
celltype_OI <- c("cholangiocyte","endothelial","hepatocyte","kupffer","monocyte","neutrophil","stellate")
hepa_celltypeOI <- subset(x = hepa_rds, idents = celltype_OI)

## Factorize experiment timepoint
experiments <- c("0_phase_1", "3_phase_2", "6_phase_2", "12_phase_2", "24_phase_2", "36_phase_2", "48_phase_1", "48_phase_2", "72_phase_2", "96_phase_2", "120_phase_2", "168_phase_2")
hours <- c(0, 3, 6, 12, 24, 36, 48, 48, 72, 96, 120, 168)
hepa_celltypeOI@meta.data$time_experiment <- factor(hepa_celltypeOI@meta.data$time_experiment, levels = experiments)
```

```{r}
hepa_sample_info <- hepa_celltypeOI@meta.data
hepa_expression <- t(hepa_celltypeOI@assays$RNA@counts)
hepa_expression[1:5, 1:5]
```

```{r}
## change ENSEMBL to gene symbol in column names

library(biomaRt)
ensembl <- useEnsembl(biomart="ensembl", dataset = "mmusculus_gene_ensembl")
IDtoSymbol_matrix <- getBM(attributes=c("ensembl_gene_id","external_gene_name"), filters = "ensembl_gene_id", values = unique(colnames(hepa_expression)), mart = ensembl)
```

```{r}
expression_colnames <- data.frame(colnames(hepa_expression))
expression_colnames["order"] <- c(1:dim(expression_colnames)[1])
symbol_list <- merge(expression_colnames, IDtoSymbol_matrix,by.x = "colnames.hepa_expression.", by.y = "ensembl_gene_id", all.x = TRUE)

## Fill rest of NA with ENSEMBL ID (ID no longer in the ENSEMBL database)
symbol_list$external_gene_name[is.na(symbol_list$external_gene_name)] <- symbol_list$colnames.hepa_expression.[is.na(symbol_list$external_gene_name)]
symbol_list <- symbol_list[order(symbol_list$order),]

## In old version, the gene names are in uppercase characters, while the new version is not.
### New version
colnames(hepa_expression) <- symbol_list$external_gene_name

### Old version
#colnames(hepa_expression) <- casefold(symbol_list$external_gene_name, upper=TRUE)
#hepa_cluster$Symbol <- casefold(hepa_cluster$Symbol, upper=TRUE)
```


# 2. Define expressed genes

sender cell type, receiver cell type. We will analyze 7x7x11x11x3 different combinations
7 cell types ("cholangiocyte","endothelial","hepatocyte","kupffer","monocyte","neutrophil","stellate")
at 12_phase_2 for both
consider a gene to be expressed when it is expressed in at least 10% of cells in one cluster
```{r}
sender_celltype <- "hepatocyte"
receiver_celltype <- "hepatocyte"
sender_timepoint <- "0_phase_1"
receiver_timepoint <- "12_phase_2"
cluster_number <- 1

# Extract sender & receiver cells
sender_ids <- row.names(hepa_sample_info[which(hepa_sample_info$celltype == sender_celltype & hepa_sample_info$time_experiment == sender_timepoint),])
receiver_ids <- row.names(hepa_sample_info[which(hepa_sample_info$celltype == receiver_celltype & hepa_sample_info$time_experiment == receiver_timepoint),])

# Define expressed genes
expressed_genes_sender <- colSums(hepa_expression[sender_ids,] != 0) %>% sapply(function(x){x/length(sender_ids)*100}) %>% .[. >= 10] %>% names()
expressed_genes_receiver <- colSums(hepa_expression[receiver_ids,] != 0) %>% sapply(function(x){x/length(sender_ids)*100}) %>% .[. >= 10] %>% names()
```

# 3. Define gene set of interest and background of genes
This step highly depend on the research question.
Gene set of interest are the genes in the target cell population that are potentially affected by ligands expressed by signaling cells.In this research, gene set of interest is genes that are showing similar temporal trend in the receiver cell type. Background expressed genes in this research are all genes from the receiver cell type. 
```{r}
geneset_oi_raw <- hepa_cluster$Symbol[hepa_cluster$celltype == receiver_celltype & hepa_cluster$cluster_number == cluster_number]
geneset_oi <- hepa_cluster$Symbol[hepa_cluster$celltype == receiver_celltype & hepa_cluster$cluster_number == cluster_number] %>% .[. %in% rownames(ligand_target_matrix)] # only consider genes that are present in the NicheNet model

# Disappeared genes
geneset_oi_raw[!geneset_oi_raw %in% geneset_oi]

background_expressed_genes <- expressed_genes_receiver %>% .[. %in% rownames(ligand_target_matrix)]
```

# 4. Define a set of potential ligands
Putative ligand-receptor links were gathered from NicheNet's ligand-receptor data sources.
```{r}
lr_network = readRDS(url("https://zenodo.org/record/5884439/files/lr_network_mouse_21122021.rds")) # new version
#lr_network = readRDS(url("https://zenodo.org/record/3260758/files/lr_network.rds")) # old version

ligands = lr_network %>% pull(from) %>% unique()
expressed_ligands = intersect(ligands,expressed_genes_sender)

receptors = lr_network %>% pull(to) %>% unique()
expressed_receptors = intersect(receptors,expressed_genes_receiver)

lr_network_expressed = lr_network %>% filter(from %in% expressed_ligands & to %in% expressed_receptors) 
head(lr_network_expressed)
```

```{r}
potential_ligands = lr_network_expressed %>% pull(from) %>% unique()
head(potential_ligands)
```

# 5. Perform NicheNet's ligand activity analysis on the gene set of interest
predict ligand activities = ranks ligands according to how well their prior target gene predictions correspond to the observed gene expression changes. In other words, assess how well each sender_cell-ligand can predict the receiver_cell gene set compared to the background of expressed genes (predict whether a gene belongs to the receiver_cell program or not)

The true ligand with which the cells were treated should predict the differential expression observed in a dataset better than the other ligands. In other words, informative feature importance scores (=measure for ligand activity) should be higher for this ligand (genes of interest) than for others(background genes). The feature importance measures were calculated for each ligand separately.

Next, evaluate how well the feature importance scores of each dataset-ligand combination can predict the correct ligand activity state. Binary classification. For a specific ligand-dataset combination, classified "active" if the dataset contains expression data of cells treated with that ligand. Then classification evaluatino metrics (AUROC, AUPR, and PCC) are calculated. 

**Compare this result with random prediction**
```{r}
# Prediction (not ranked yet)
ligand_activities = predict_ligand_activities(geneset = geneset_oi, background_expressed_genes = background_expressed_genes, ligand_target_matrix = ligand_target_matrix, potential_ligands = potential_ligands)

# Rank the prediction
ligand_activities %>% arrange(-aupr)
best_upstream_ligands = ligand_activities %>% top_n(20, aupr) %>% arrange(-aupr) %>% pull(test_ligand) %>% unique()
```

```{r}
#write.table(ligand_activities, file="C:/Users/pc/Desktop/MastersProject/Dynamic_NicheNet/Ligand_Activity_Prediction_AUPR_NicheNet_v2.csv", sep = ",")
```


```{r}
# show histogram of ligand activity scores
p_hist_lig_activity = ggplot(ligand_activities, aes(x=aupr)) + 
  geom_histogram(color="black", fill="darkorange")  + 
  # geom_density(alpha=.1, fill="orange") +
  geom_vline(aes(xintercept=min(ligand_activities %>% top_n(20, aupr) %>% pull(aupr))), color="red", linetype="dashed", size=1) + 
  labs(x="ligand activity (AUPR)", y = "# ligands") +
  theme_classic()
p_hist_lig_activity

```

# 6. Infer target genes of top-ranked ligands and visualize in a heatmap

## 1) infer target genes
If you consider more than the top 250 targets, you will infer more, but less confident ligand-target links.

*drop_na() : NA value in target and weight column means that there are no DE gene in the top 250 targets (gene set of interest). In other words, there were potential targets in top250 (geneset of interest) to be regulated by the ligand.
```{r}
active_ligand_target_links_df = best_upstream_ligands %>% lapply(get_weighted_ligand_target_links, geneset = geneset_oi, ligand_target_matrix = ligand_target_matrix, n = 250) %>% bind_rows() %>% drop_na()

nrow(active_ligand_target_links_df)
head(active_ligand_target_links_df)

```

## 2) Set cutoff for visualization
Lowering the quantile cutoff (0.25) will result in a more dense heatmap, whereas highering this cutoff will result in a more sparse heatmap.
```{r}
# visualization
active_ligand_target_links = prepare_ligand_target_visualization(ligand_target_df = active_ligand_target_links_df, ligand_target_matrix = ligand_target_matrix, cutoff = 0.25) # score below threshold will be set as 0

head(active_ligand_target_links_df)
```

## 3) Plot putatively active ligand-target links
The order of the ligands accord to the ranking of the ligand activity prediction.
```{r}
order_ligands = intersect(best_upstream_ligands, colnames(active_ligand_target_links)) %>% rev()
order_targets = active_ligand_target_links_df$target %>% unique()
vis_ligand_target = active_ligand_target_links[order_targets,order_ligands] %>% t()

p_ligand_target_network = vis_ligand_target %>% make_heatmap_ggplot("Prioritized ligands","Predicted target genes", color = "purple",legend_position = "top", x_axis_position = "top",legend_title = "Regulatory potential") + scale_fill_gradient2(low = "whitesmoke",  high = "purple", breaks = c(0,0.05,0.10)) + theme(axis.text.x = element_text(face = "italic"))

p_ligand_target_network

```

# 7. Validation

## 1) Ligand-receptor network inference for top-ranked ligands
Look at which receptors of the receiver cell population can potentially bind to the prioritized ligands from the sender cell population
```{r}
# get the ligand-receptor network of the top-ranked ligands
lr_network_top = lr_network %>% filter(from %in% best_upstream_ligands & to %in% expressed_receptors) %>% distinct(from,to)
best_upstream_receptors = lr_network_top %>% pull(to) %>% unique()

# get the weights of the ligand-receptor interactions as used in the NicheNet model
weighted_networks = readRDS(url("https://zenodo.org/record/5884439/files/weighted_networks_nsga2r_final_mouse.rds")) # New version
#weighted_networks = readRDS(url("https://zenodo.org/record/3260758/files/weighted_networks.rds")) # Old version
lr_network_top_df = weighted_networks$lr_sig %>% filter(from %in% best_upstream_ligands & to %in% best_upstream_receptors)

# convert to a matrix
lr_network_top_df = lr_network_top_df %>% spread("from","weight",fill = 0)
lr_network_top_matrix = lr_network_top_df %>% dplyr::select(-to) %>% as.matrix() %>% magrittr::set_rownames(lr_network_top_df$to)

# perform hierarchical clustering to order the ligands and receptors
# hierarchical clustering groups similar objects into clusters
# Ward.D2 = Ward algorithm = creates groups such that variance is minimized within clusters. Dissimilarities are squared before clustering.
dist_receptors = dist(lr_network_top_matrix, method = "binary")
hclust_receptors = hclust(dist_receptors, method = "ward.D2")
order_receptors = hclust_receptors$labels[hclust_receptors$order]

dist_ligands = dist(lr_network_top_matrix %>% t(), method = "binary")
hclust_ligands = hclust(dist_ligands, method = "ward.D2")
order_ligands_receptor = hclust_ligands$labels[hclust_ligands$order]

```

```{r}
vis_ligand_receptor_network = lr_network_top_matrix[order_receptors, order_ligands_receptor]
p_ligand_receptor_network = vis_ligand_receptor_network %>% t() %>% make_heatmap_ggplot("Prioritized ligands","Receptors", color = "mediumvioletred", x_axis_position = "top",legend_title = "Prior interaction potential")
p_ligand_receptor_network

```

## 2) Visualize expression of top-predicted ligands and their target genes in a combined heatmap
Further look at expression of ligands & target genes. 
Combine plots of ligand activity, ligand expression, target gene expression, and ligand-target regulatory potential. 
```{r}
library(RColorBrewer)
library(cowplot)
library(ggpubr)

```
**ligand activity matrix**
```{r}
ligand_aupr_matrix = ligand_activities %>% dplyr::select(aupr) %>% as.matrix() %>% magrittr::set_rownames(ligand_activities$test_ligand)

vis_ligand_aupr = ligand_aupr_matrix[order_ligands, ] %>% as.matrix(ncol = 1) %>% magrittr::set_colnames("AUPR")

```

```{r}
p_ligand_aupr = vis_ligand_aupr %>% make_heatmap_ggplot("Prioritized ligands","Ligand activity", color = "darkorange",legend_position = "top", x_axis_position = "top", legend_title = "AUPR\ntarget gene prediction ability)")
p_ligand_aupr

```
# 8. Gene Interpretation

## 1) Temporal trends
```{r}
logCPM.obs = read.csv("C:/Users/pc/Desktop/MastersProject/Dynamic_NicheNet/DE_geneID+celltype_logCPM-obs.csv")
logCPM.fit = read.csv("C:/Users/pc/Desktop/MastersProject/Dynamic_NicheNet/DE_geneID+celltype_logCPM-fit.csv")
```

```{r}
plot_trend <- function(interesting_genes, celltype){
  par(mfrow=c(2,3))
  
  for(gene in interesting_genes) {
    gene_celltype <- paste0(gene, "_", celltype)
    logCPM.obs.i <- logCPM.obs[gene_celltype,]
    logCPM.fit.i <- logCPM.fit[gene_celltype,]
    plot(hours, logCPM.obs.i, ylab="log-CPM", main=gene_celltype, pch=16)
    lines(hours, logCPM.fit.i, col="red", lwd=2)
  }
}
```


**Active Ligands**
```{r}
plot_trend(order_ligands, sender_celltype)
```

**Active Targets**
```{r}
plot_trend(order_targets, receiver_celltype)
```

**Receptors**
```{r}
plot_trend(order_receptors, receiver_celltype)
```

## 2) GO terms
Ont = Ontology that the GO term belongs to. BP (Biological Pathway), CC(Cellular Component), MF(Molecular Function)
N = number of genes in the GO term
DE = number of genes in the "DE" set
P.DE = p-value for over-representation of the GO term in the set.

```{r}
# GO <- goana(order_receptors, species = "Mm")
# topgo <- topGO(GO)
```





Trial
```{r}
# tab = read.csv("C:/Users/pc/Desktop/MastersProject/Dynamic_NicheNet/tab.csv")
# tab["de"] <- rep(0, dim(tab)[1])
```

```{r}
# GO_genes <- logCPM.fit[which(rownames(logCPM.fit) %in% paste0(order_receptors, "_", receiver_celltype)),]
# tab$de[which(rownames(tab) %in% paste0(order_receptors, "_", receiver_celltype))] <- 1

```

```{r}
# library(topGO)
# gene_universe <- tab$de
# gene_universe <- factor(gene_universe)
# names(gene_universe) <- rownames(tab)
# head(gene_universe)

```
```{r}
# go_data <- new("topGOdata",
#                ontology = "BP",
#                allGenes = gene_universe,
#                nodeSize = 5,
#                annotationFun = annFUN.org,
#                mapping = "org.Mm.eg",
#                ID = "symbol")
```





