---
title: "DE_pseudobulk"
author: "Heesoo Song"
date: '2021 12 1 '
output: html_document
---

Extension of *DE_EdgeR.Rmd*. Here, we will create aggregate matrix to observe change spanning over celltype rather than time point.
https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf

```{r}
#install.packages('Matrix.utils')
#install.packages("statmod")
#BiocManager::install("org.Mm.eg.db")
```


```{r, message = FALSE}
library(SeuratObject)
library(Seurat)
library(nichenetr)
library(tidyverse)
library("ggplot2")
library(muscat)
library(SingleCellExperiment)
library(purrr)
library(Matrix.utils)
library(RColorBrewer)
library(pheatmap)
library(cluster)
```

```{r}
# Load processed expression data of interacting cells
hepatectomy_expression = readRDS("C:/Users/pc/Desktop/MastersProject/Dynamic_NicheNet/seurat_obj_phx_25102021.rds")

# Total 91021 cells, with 5000 genes
```

```{r}
celltype_OI <- c("cholangiocyte","endothelial","hepatocyte","kupffer","monocyte","neutrophil","stellate")
```

```{r}
hepa_celltypeOI = subset(x = hepatectomy_expression, idents = celltype_OI)

experiments <- c("0_phase_1", "3_phase_2", "6_phase_2", "12_phase_2", "24_phase_2", "36_phase_2", "48_phase_1", "48_phase_2", "72_phase_2", "96_phase_2", "120_phase_2", "168_phase_2")

hepa_celltypeOI@meta.data$time_experiment <- factor(hepa_celltypeOI@meta.data$time_experiment, levels = experiments)
```


# Create Aggregate Count Matrix

```{r}
# Extract raw counts and metadata to create SingleCellExperiment object
counts <- hepa_celltypeOI@assays$RNA@counts
metadata <- hepa_celltypeOI@meta.data

# Set up metadata as desired for aggregation and DE analysis
metadata$cluster_id <- factor(hepa_celltypeOI@active.ident)
#metadata$cluster_sample <- paste0(factor(metadata$cluster_id), ".", factor(metadata$time_experiment)) # This line is added

# Create single cell experiment object
sce <- SingleCellExperiment(assays = list(counts = counts),
                            colData = metadata)

```


```{r}
#Named vector of cluster names
kids <- purrr::set_names(levels(sce$cluster_id))
kids

#Total number of clusters
nk <- length(kids)
nk

#Named vector of sample names
sids <- purrr::set_names(levels(factor(sce$time_experiment)))
#sids <- purrr::set_names(levels(factor(sce$cluster_sample)))
sids[1:10]

#Total number of samples
ns <- length(sids)
ns
```

```{r}
# Generate sample level metadata

## Determine the number of cells per sample
table(sce$time_experiment)

## Turn named vector into a numeric vector of cells per sample
n_cells <- as.numeric(table(sce$time_experiment))

## Determine how to reorder the samples (rows) of the metadata to match the order of sample names in sids vector
m <- match(sids, sce$time_experiment)

```

```{r}
# Identify groups for aggregation of counts
groups <- colData(sce)[, c("cluster_id", "time_experiment")]
```

```{r}
counts(sce)[1:6,1:6]
```


```{r}
# Create aggregate matrix
pb <- aggregate.Matrix(t(counts(sce)),
                       groupings = groups, fun = "sum")

class(pb)
dim(pb)
pb[1:6, 1:6]
```

```{r}
# Not every cluster is present in all samples; create a vector that represents how to split samples
# But actually we do have all cell types! :) Maybe skip this step
splitf <- sapply(stringr::str_split(rownames(pb), pattern="_", n=2), '[', 1)
```

```{r}
# Turn into a list and split the list into components for each cluster and transform, so rows are genes and columns are samples and make rownames as the sample IDs
library(magrittr)
pb <- split.data.frame(pb, 
                       factor(splitf)) %>%
  lapply(function(u) 
    set_colnames(t(u), 
                 stringr::str_extract(rownames(u), "(?<=_)[:alnum:]+_phase_[0-9]"))) 

class(pb)

# Explore the different components of list
names(pb)
```

```{r}
pb[[1]][1:6, 1:6]
```

```{r}
#Change row names into geneID + celltype
for (i in 1:length(names(pb))){
  rownames(pb[[i]]) <- paste0(rownames(pb[[i]]), ".", names(pb)[i])
}

pb[[1]][1:6, 1:6]
```


```{r}
# Combine matrices
pb_combined <- pb[[1]]
dim(pb_combined)

for (i in 2:length(names(pb))) {
  pb_combined <- rbind(pb_combined, pb[[i]])
}
dim(pb_combined)
pb_combined[1:6, 1:12]
```


```{r}
# Print out the table of cells in each cluster-sample group
options(width = 100)
table(sce$cluster_id, sce$time_experiment)
```

# 2. Differential gene expression
The aim is to examine the relationship between gene abundances and time points per each cell-type. In other words, we are looking for genes that change expression level over time in a time course experiment. Given enough number of time points, it is reasonable to assume smooth changes in gene expression. In this way, polynomial or a cubic spline curve can be used to model gene expression along time with a certain number of degrees of freedom.


Perform DE analysis
```{r}
cluster_counts <- as.data.frame(as.matrix(pb_combined))
colnames(cluster_counts)
```

Explore aggregated count data by MDS (multi-dimensional scaling) plots. This visualizes the differences between the expression profiles of different samples (cholangiocyte) in two dimensions.
```{r}
library(edgeR)
plotMDS(cluster_counts, labels=colnames(cluster_counts))
```


```{r}
hours <- c(0, 3, 6, 12, 24, 36, 48, 48, 72, 96, 120, 168)
# X <- poly(hours, degree=3)
# design <- model.matrix(~X)
# design
```

Use cubic regression spline curve with 3 degrees of freedom. The advantage of using a cubic spline curve is that it provides more stable fit at the end points compared to a polynomial. The spline curve with 3 degrees of freedom has 2 knots where cucbic polynomicals are splined together.
```{r}
library(splines)
X <- ns(hours, df = 4)
design <- model.matrix(~X)
design
```
Estimate dispersion. The negative binomial (NB) dispersion is estimated. This returns the **DGEList** object with additional entries for the estimated NB dispersion for each gene.
```{r}
y <- DGEList(counts = cluster_counts, group = hours)

# filter out lowly expressed genes
#keep <- filterByExpr(y)
#y <- y[keep, , keep.lib.sizes = FALSE]

# TMM Normalisation
#y <- calcNormFactors(y)
#y$samples
```

```{r}
y <- estimateDisp(y, design) # reduced from 134596 genes -> 92949 genes
sqrt(y$common.dispersion)

# plot the tagwise dispersion against log2-CPM (counts per million)
plotBCV(y)
```


Fit a quasi-likelihood NB generalized log-linear model
```{r}
fit <- glmQLFit(y, design, robust=TRUE)
plotQLDisp(fit)
```

Fit a quasi-likelihood negative binomial generalized log-linear model to count data.
```{r}
fit <- glmQLFTest(fit, coef=2:5)
```

Multiple testing across genes and contrasts. The total number of genes with significant (5% FDR) changes at different time points. 
All spline coefficients are tested together. Thus t-tests for the individual coefficients and the logFC values do not have any interpretable meaning.
```{r}
sig_num <- summary(decideTests(fit))
sig_num
# Before filtering: NotSig = 119186, Sig = 15410
# After filtering: NotSig = 77693, Sig = 15256
```
#NEW
```{r}
# Genes with most significant time effects
tab <- as.data.frame(topTags(fit, n=90000)) %>% rownames_to_column(var = "gene")
tab["order"] <- c(1:length(rownames(tab)))
head(tab)
```

LogFC = log fold change, log difference between groups.
LogCPM = log count per million, measure of expression level.

```{r}
extract_geneID <- as.data.frame(str_split(tab$gene, "[.]", simplify=TRUE))
tab$ENSEMBL <- extract_geneID$V1
tab$celltype <- extract_geneID$V2
```

Change ENSEMBL gene IDs to gene symbols

```{r}
library(biomaRt)
ensembl <- useEnsembl(biomart="ensembl", dataset = "mmusculus_gene_ensembl")
IDtoSymbol_matrix <- getBM(attributes=c("ensembl_gene_id","external_gene_name"), filters = "ensembl_gene_id", values = unique(extract_geneID$V1), mart = ensembl)
```

```{r}
tab <- merge(tab, IDtoSymbol_matrix, by.x="ENSEMBL", by.y = "ensembl_gene_id", all.x=TRUE)
colnames(tab)[length(colnames(tab))] <- "Symbol"

# Fill rest of NA with ENSEMBL ID (ID no longer in the ENSEMBL database)
tab$Symbol[which(is.na(tab$Symbol))] <- tab$ENSEMBL[which(is.na(tab$Symbol))]

# New column with combination of gene symbol & celltype
tab$Symbol_celltype <- paste0(tab$Symbol, "_", tab$celltype)

tab <- tab %>% column_to_rownames(var = "gene")

# reorder column and row
tab <- tab[, c(2,3,4,5,6,7,8,9,10,11,1,12)]
tab <- tab[order(tab$order),]
head(tab)
```


```{r}
#write.table(tab, file="C:/Users/pc/Desktop/MastersProject/Dynamic_NicheNet/DE_geneID+celltype_Filter_5param.csv", sep = ",")
```


```{r}
logCPM.obs <- cpm(y, log=TRUE, prior.count=fit$prior.count)
logCPM.fit <- cpm(fit, log=TRUE)
```

```{r}
#write.table(logCPM.obs, file="C:/Users/pc/Desktop/MastersProject/Dynamic_NicheNet/DE_geneID+celltype_logCPM-obs.csv", sep = ",")
#write.table(logCPM.fit, file="C:/Users/pc/Desktop/MastersProject/Dynamic_NicheNet/DE_geneID+celltype_logCPM-fit.csv", sep = ",")

```

# 3. Examine significant genes
## 3-1. Cluster Analysis


```{r}
# Clustering based on fitted values
normalized_counts <- logCPM.fit

# Extract normalized counts for only the significant genes
sig_norm <- data.frame(normalized_counts) %>%
  rownames_to_column(var = "gene") %>%
  dplyr::filter(gene %in% rownames(tab)[1:sig_num[2]])

colnames(sig_norm) <- c("gene", experiments)
sig_norm <- sig_norm %>% column_to_rownames(var="gene")
head(sig_norm)
```


### Calculate distance
```{r}
library("amap")
gene_dist <- Dist(sig_norm[, 1:length(colnames(sig_norm))], method="spearman")
gene_hclust <- hclust(gene_dist, method = "ward.D2")
```

### Find the best cluster number k

**Dendrogram**
The height of the dendrogram indicates the order in which the clusters were joined. A more informative dendrogram can be created where the heights reflect the distance between the clusters.
```{r}
plot(gene_hclust, labels=FALSE, hang=-1)
```


**Silhouette coefficient**
```{r}
# Determine the best number of clusters
sil_coef_vector <- c()
for (num_cluster in 3:60){
  sil_cutree <- cutree(gene_hclust, k = num_cluster)
  sil_cl <- silhouette(sil_cutree ,gene_dist)
  sil_cl_summary <- summary(sil_cl)
  sil_coefficient <- as.vector(sil_cl_summary$si.summary["Mean"])
  sil_coef_vector<- c(sil_coef_vector, sil_coefficient)
}
#summary(sil_cl)
```

```{r}
plot(3:60,sil_coef_vector,"b", xlab="k", ylab="silhouette coefficient")
```

**C-index**
```{r}
# library("clusterSim")
# # Determine the best number of clusters
# c_index_vector <- c()
# for (num_cluster in 3:20){
#   c_cutree <- cutree(gene_hclust, k = num_cluster)
#   c_cl <- index.C(gene_dist, c_cutree)
#   c_index_vector<- c(c_index_vector, c_cl)
# }
```

```{r}
#plot(3:20,c_index_vector,"b", xlab="k", ylab="C-index")
```

### Clustering
```{r}
gene_cluster <- data.frame(cutree(gene_hclust, k = 4)) %>%
  rownames_to_column()
colnames(gene_cluster) <- c("gene", "cluster_number")
head(gene_cluster)
```

### Plot clusters
```{r}
par(mfrow=c(2,2))

histogram <- c()
for(i in 1:4) {
  cluster_genes <- gene_cluster[gene_cluster$cluster_number == i, "gene"]
  logCPM.fit.i <- data.frame(logCPM.fit[cluster_genes,])
  colnames(logCPM.fit.i) <- hours
  histogram <- c(histogram, length(cluster_genes))
  
  
  # Commented line below shows trend lind of each genes
  # plot(hours, logCPM.fit.i[1,], type = "l", ylab="log-CPM", main=paste0("cluster ", i, ": ", length(cluster_genes), " genes"), pch=16)
  # for (j in 2:10){#length(cluster_genes)) {
  #    lines(hours, logCPM.fit.i[j,])
  # }
  
  # Box plot of clusters through timepoint
  box <- boxplot(logCPM.fit.i, ylim=c(-15, 15), ylab="log-CPM", main=paste0("cluster ", i, ": ", length(cluster_genes), " genes"))
  boxaverage <- box$stats[3,]
  plot(hours, boxaverage, type="l", ylab="log-CPM", main=paste0("cluster ", i, ": ", length(cluster_genes), " genes"), pch=16)
  
}


#histogram <- as.dataframe(histogram)
```

```{r}
plot(histogram, type="h", lwd=10, ylim=c(0, 6000), xlab="Cluster", ylab="Cluster Size")
text(histogram, format(histogram), pos=3, col="red", cex=0.8)
```



### Save cluster number as csv
```{r}
cluster_number_merged <- merge(x = tab[1:sig_num[2],], y = gene_cluster, by.x = 0 , by.y = "gene", all = TRUE)
head(cluster_number_merged)
```

```{r}
#write.table(cluster_number_merged, file="C:/Users/pc/Desktop/MastersProject/Dynamic_NicheNet/DE_pseudobulk_cluster.csv", sep = ",")
```

## 3-2. Common genes across cell types
```{r}
# Create false dataframe
common_genes <- as.data.frame(matrix(data = FALSE, nrow = length(levels(factor(cluster_number_merged$ENSEMBL))), ncol=7))
colnames(common_genes) <- celltype_OI
rownames(common_genes) <- as.vector(levels(factor(cluster_number_merged$ENSEMBL)))

# Change False into True if it exists in 
for (n in 1:dim(cluster_number_merged)[1]) {
  common_genes[cluster_number_merged$ENSEMBL[n], cluster_number_merged$celltype[n]] <- TRUE
}

common_genes["count"] <- rowSums(common_genes)

head(common_genes)
```

```{r, message=FALSE}
gene_symbol <- c()
for (n in 1:dim(common_genes)[1]) {
  gene_symbol[n] <- tab$Symbol[which(tab$ENSEMBL == rownames(common_genes)[n])][1]
}

common_genes["symbol"] <- gene_symbol

head(common_genes)
```

```{r}
#write.table(common_genes, file="C:/Users/pc/Desktop/MastersProject/Dynamic_NicheNet/Dynamic_NicheNet/DE_pseudobulk_commonGenes.csv", sep = ",")
```

```{r}
common_genes[common_genes$count == 7, "symbol"]
```

## 3-3. Heatmap

Heatmap
- explore clustering of the significant genes

```{r}
normalized_counts <- logCPM.obs

# Extract normalized counts for only the significant genes
sig_norm_heatmap <- data.frame(normalized_counts) %>%
  rownames_to_column(var = "gene") %>%
  dplyr::filter(gene %in% rownames(tab)[1:20])

colnames(sig_norm_heatmap) <- c("gene", experiments)


# Set up row names
gene_symbol <- c()
for (n in 1:length(sig_norm_heatmap$gene)) {
  gene_symbol[n] <- tab$Symbol_celltype[which(rownames(tab) == sig_norm_heatmap$gene[n])]
}

rownames(sig_norm_heatmap) <- gene_symbol
```

```{r}
# Set a color palette
heat_colors <- brewer.pal(20, "RdYlBu")

# Run pheatmap using the metadata for the annotation
pheatmap(sig_norm_heatmap[, 2:length(colnames(sig_norm_heatmap))],
         color = heat_colors,
         cluster_rows = T,
         cluster_cols = F, # Make this TRUE if you want to look cluster between different time points
         show_rownames = T,
         border_color = NA,
         fontsize = 10,
         scale = "row",
         fontsize_row = 10,
         height = 20,
         main = paste0("Top DE gene clusters"))
```

Before filtering:  
"Dlgap5_endothelial", "Plk1_endothelial", "Cdc20_endothelial", "Kif18b_endothelial" are clustered together in the heatmap.  

After filtering:  
"Dlgap5_endothelial", "Plk1_endothelial", "Knstrn_endothelial", "Kif18b_endothelial" are clustered together. But still, Cdc20_endothelial exist in the significant gene list.

Dlgap5  
  - DLG Associated Protein 5  
  - GO: Phosphoprotein phosphatase activity  
  - Potential cell cycle regulator that may play a role in carcinogenesis of cancer cells  
  - Disease associated: Hepatocellular carcinoma  
  
Plk1  
  - Polo Like Kinase 1  
  - GO: transferase activity, transferring phosphorus-containing groups, protein tyrosine kinase activity  
  - Highly expressed during mitosis  
  
Cdc20  
  - Cell Division Cycle 20  
  - GO: enzyme binding, protein C-terminus binding  
  - Regulatory protein interacting with several other proteins at multiple points in the cell cycle  
  
Kif18b  
  - Kinesin Family Member 18B  
  - GO: ATP hydrolysis activity, microtubule motor activity  
  - In complex with Kif2c, constitutes the major microtubule plus-end depolymerizing activity in mitotic cells.   

## 3-4. Plot important individual genes

```{r}
# par(mfrow=c(2,3))
# #interesting_genes <- c("Dlgap5_endothelial", "Plk1_endothelial", "Cdc20_endothelial", "Kif18b_endothelial")
# #interesting_genes <- c("Dlgap5_endothelial", "Plk1_endothelial", "Knstrn_endothelial", "Cdc20_endothelial", "Kif18b_endothelial")
# #interesting_genes <- c("Osm_neutrophil", "Il1r1_hepatocyte","Csf1_neutrophil", "Cxcl1_stellate")
# interesting_genes <- c("Il6_kupffer", "Tnf_kupffer", "Csf1_stellate","Cxcl1_hepatocyte", "Insr_endothelial", "Bmp10_stellate")
# 
# FlybaseID_list = c()
# for(i in 1:length(interesting_genes)) {
#   FlybaseID <- rownames(tab)[which(tab$Symbol_celltype == interesting_genes[i])]
#   Symbol <- interesting_genes[i]
#   logCPM.obs.i <- logCPM.obs[FlybaseID,]
#   logCPM.fit.i <- logCPM.fit[FlybaseID,]
#   plot(hours, logCPM.obs.i, ylab="log-CPM", main=Symbol, pch=16)
#   lines(hours, logCPM.fit.i, col="red", lwd=2)
#   FlybaseID_list[i] <- FlybaseID
# }
```

```{r}
# FlybaseID_list
```

## 3-5. Violin Plot
Violin Plot
- visualize single cell expression distributions in each cluster
```{r}
# for (i in 1:length(FlybaseID_list)) {
#   feature_OI <- str_split(FlybaseID_list[i], "[.]") %>% unlist()
#   
#   hepa_clusterOI <- subset(x = hepa_celltypeOI, idents = feature_OI[2])
#   
#   print(VlnPlot(hepa_celltypeOI, group.by = "time_experiment", features = feature_OI[1], slot = "counts", pt.size = 0))
# }
```

##3-6. UMAP & Feature Plot

Plot interesting genes back in UMAP and Feature plot
```{r}
# Plot_features <- function(gene, cluster_OI) {
# 
#   for (i in 1:12){
#     genename <- tab[which(rownames(tab) == paste0(gene, ".", cluster_OI)),"Symbol"]
#     
#     dir.create(paste0("./FeaturePlots/", cluster_OI, "_", genename))
#     jpeg(paste0("./FeaturePlots/", cluster_OI, "_", genename, "/",cluster_OI, "_", genename, "-", experiments[i], ".jpg"), width = 1500, height = 750)
#     
#     subset <- subset(x = hepa_celltypeOI, subset = time_experiment == experiments[i])
#     highlight_cells <- Cells(subset)
#     dimplot <- DimPlot(hepa_celltypeOI, cells.highlight= list(highlight_cells), sizes.highlight = 0.5, label = TRUE, label.size=3)
#   
#     feature_baseplot <- FeaturePlot(subset, features = gene, min.cutoff = "q10", max.cutoff = "q90" ,label = TRUE)
#     featureplotwithtitle <- feature_baseplot + labs(title = paste0(genename, ": ",experiments[i]))
#     
#     print(dimplot + featureplotwithtitle)
#     
#     dev.off()
#   }
# }
```

```{r, message=FALSE}
# for (i in 1:length(FlybaseID_list)) {
#   feature_OI <- str_split(FlybaseID_list[i], "[.]") %>% unlist()
#   
#   hepa_clusterOI <- subset(x = hepa_celltypeOI, idents = feature_OI[2])
#   
#   Plot_features(feature_OI[1], feature_OI[2])
# }
```

